### PR TITLE
Magic phase 8 - spellcasting action and activity

### PR DESCRIPTION
**Pull request by KorGgenT**
#### Summary

```SUMMARY: Infrastructure "Add spellcasting activity into code"```
#### Purpose of change

This is the penultimate phase of the magic series of PRs -- spellcasting. This adds the spellcasting action via handle_action.cpp and the spellcasting activity in activity_handlers. This also adds two options in the debug menu and the mana panel.

I would like some input on how to hide the mana panel if the external option is not set -- i have tried just checking it in the init function, but external options are not loaded at that point. Maybe hiding the spell debug functions if there are no spells would be prudent as well.
EDIT: I have come up with solutions to both of the above issues.

This is the last PR that will contain only C++ code before I will PR the actual mod.

### CODE REVIEW

File: src/handle_action.cpp [hunk 1]
```diff
@@ -1251,6 +1252,100 @@ static void open_movement_mode_menu()
...
+    for( spell_id sp : spells ) {
+        spell temp_spell = u.magic.get_spell( sp );
+        std::string nm = temp_spell.name();
+        uilist_entry entry( nm );
+        if( temp_spell.can_cast( u ) ) {
+            can_cast_spells = true;
+        } else {
+            entry.enabled = false;
+        }
+        std::string turns = temp_spell.casting_time() >= 100 ? string_format( "%i turns",
```
Comment (AMurkin) [inline src/handle_action.cpp]:
```suggestion
        std::string turns = temp_spell.casting_time() >= 100 ? string_format( _("%i turns" ),
```

File: src/handle_action.cpp [hunk 2]
```diff
@@ -1251,6 +1252,100 @@ static void open_movement_mode_menu()
...
+                            temp_spell.casting_time() / 100 ) : string_format( "%i moves", temp_spell.casting_time() );
```
Comment (AMurkin) [inline src/handle_action.cpp]:
```suggestion
                            temp_spell.casting_time() / 100 ) : string_format( _( "%i moves" ), temp_spell.casting_time() );
```

File: src/handle_action.cpp [hunk 3]
```diff
@@ -1251,6 +1252,100 @@ static void open_movement_mode_menu()
...
+            case none_energy:
+                cost = colorize( _( "none" ), c_light_gray );
+                break;
+            default:
+                debugmsg( "ERROR: %s has invalid energy_type", temp_spell.id().c_str() );
+                break;
+        }
+        entry.ctxt = string_format( "%3i (%3s) %3i %3i %% %10s %4s %3i", temp_spell.get_level(),
+                                    temp_spell.is_max_level() ? _( "MAX" ) : temp_spell.exp_progress(), temp_spell.range(),
+                                    static_cast<int>( round( 100.0f * temp_spell.spell_fail( u ) ) ), _( turns ), cost,
```
Comment (AMurkin) [inline src/handle_action.cpp]:
```suggestion
                                    static_cast<int>( round( 100.0f * temp_spell.spell_fail( u ) ) ), turns, cost,
```

File: src/iuse_actor.cpp [hunk 4]
```diff
@@ -2125,6 +2127,107 @@ ret_val<bool> musical_instrument_actor::can_use( const player &p, const item &,
...
+    }
+    std::vector<uilist_entry> uilist_initializer;
+    bool know_it_all = true;
+    for( const std::string sp_id_str : spells ) {
+        const spell_id sp_id( sp_id_str );
+        const std::string sp_nm = sp_id.obj().name;
+        uilist_entry entry( sp_nm );
+        if( p.magic.knows_spell( sp_id ) ) {
+            const spell sp = p.magic.get_spell( sp_id );
+            entry.ctxt = string_format( "Level %u", sp.get_level() );
```
Comment (AMurkin) [inline src/iuse_actor.cpp]:
```suggestion
            entry.ctxt = string_format( _( "Level %u" ), sp.get_level() );
```

File: src/debug_menu.cpp [hunk 5]
```diff
@@ -1407,7 +1418,56 @@ void debug()
...
+                                    break;
+            case DEBUG_LEARN_SPELLS:
+                if ( spell_type::get_all().empty() ) {
+                    add_msg( m_bad, _( "There are no spells to learn. You must install a mod that adds some." ) );
+                }
+                else {
+                    for ( const spell_type &learn : spell_type::get_all() ) {
+                        g->u.magic.learn_spell( &learn, g->u, true );
+                    }
+                    add_msg( m_good, _( "You have become an Archwizardpriest! What will you do with your newfound power?" ) );
```
Comment (AMurkin) [inline src/debug_menu.cpp]:
```suggestion
                    add_msg( m_good, _( "You have become an Archwizardpriest!  What will you do with your newfound power?" ) );
```

File: data/json/game_balance.json [hunk 6]
```diff
@@ -166,5 +166,12 @@
     "info": "Disable additional spawn of large groups of monsters fighting each other on swamps.",
     "stype": "bool",
     "value": false
+  },
+  {
+    "type": "EXTERNAL_OPTION",
+    "name": "MANA",
+    "info": "Mana is an energy source generated from players' bodies used for casting magic. If true, adds Mana to a new panel in the sidebar.",
```
Comment (AMurkin) [inline data/json/game_balance.json]:
```suggestion
    "info": "Mana is an energy source generated from players' bodies used for casting magic.  If true, adds Mana to a new panel in the sidebar.",
```

File: src/activity_handlers.cpp [hunk 7]
```diff
@@ -3550,3 +3555,165 @@ void activity_handlers::tree_communion_do_turn( player_activity *act, player *p
...
+        do {
+            trajectory = th.target_ui( casting );
+            if( !trajectory.empty() ) {
+                target = trajectory.back();
+                target_is_valid = casting.is_valid_target( target );
+            } else {
+                target_is_valid = false;
+            }
+            if( !target_is_valid ) {
+                if( query_yn( _( "Stop casting spell? Time spent will be lost." ) ) ) {
```
Comment (AMurkin) [inline src/activity_handlers.cpp]:
```suggestion
                if( query_yn( _( "Stop casting spell?  Time spent will be lost." ) ) ) {
```

File: src/activity_handlers.cpp [hunk 8]
```diff
@@ -3550,3 +3555,165 @@ void activity_handlers::tree_communion_do_turn( player_activity *act, player *p
...
+
+    // no turning back now. it's all said and done.
+    bool success = rng_float( 0.0f, 1.0f ) >= casting.spell_fail( *p );
+    int exp_gained = casting.casting_exp( *p );
+    if( !success ) {
+        p->add_msg_if_player( m_bad, "You lose your concentration!" );
+        if( !casting.is_max_level() ) {
+            // still get some experience for trying
+            casting.gain_exp( exp_gained / 5 );
+            p->add_msg_if_player( m_good, _( "You gain %i experience. New total %i." ), exp_gained / 5,
```
Comment (AMurkin) [inline src/activity_handlers.cpp]:
```suggestion
            p->add_msg_if_player( m_good, _( "You gain %i experience.  New total %i." ), exp_gained / 5,
```

File: src/activity_handlers.cpp [hunk 9]
```diff
@@ -3550,3 +3555,165 @@ void activity_handlers::tree_communion_do_turn( player_activity *act, player *p
...
+        case none_energy:
+        default:
+            break;
+    }
+    if( !casting.is_max_level() ) {
+        // reap the reward
+        if( casting.get_level() == 0 ) {
+            casting.gain_level();
+            p->add_msg_if_player( m_good,
+                                  _( "Something about how this spell works just clicked! You gained a level!" ) );
```
Comment (AMurkin) [inline src/activity_handlers.cpp]:
```suggestion
                                  _( "Something about how this spell works just clicked!  You gained a level!" ) );
```

File: src/activity_handlers.cpp [hunk 10]
```diff
@@ -3550,3 +3555,165 @@ void activity_handlers::tree_communion_do_turn( player_activity *act, player *p
...
+        } else {
+            casting.gain_exp( exp_gained );
+            p->add_msg_if_player( m_good, _( "You gain %i experience, New total %i." ), exp_gained,
```
Comment (AMurkin) [inline src/activity_handlers.cpp]:
```suggestion
            p->add_msg_if_player( m_good, _( "You gain %i experience.  New total %i." ), exp_gained,
```

File: src/debug_menu.cpp [hunk 11]
```diff
@@ -1407,7 +1418,56 @@ void debug()
 #endif
                 popup( popup_msg );
             }
+                                    break;
+            case DEBUG_LEARN_SPELLS:
+                if ( spell_type::get_all().empty() ) {
+                    add_msg( m_bad, _( "There are no spells to learn. You must install a mod that adds some." ) );
```
Comment (AMurkin) [inline src/debug_menu.cpp]:
```suggestion
                    add_msg( m_bad, _( "There are no spells to learn.  You must install a mod that adds some." ) );
```

File: src/magic.cpp [hunk 12]
```diff
@@ -195,3 +195,938 @@ bool spell_type::is_valid() const
...
+    }
+    if( energy_source() == hp_energy ) {
+        auto pair = get_hp_bar( energy_cost(), p.get_hp_max() / num_hp_parts );
+        return colorize( pair.first, pair.second );
+    }
+    if( energy_source() == stamina_energy ) {
+        auto pair = get_hp_bar( energy_cost(), p.get_stamina_max() );
+        return colorize( pair.first, pair.second );
+    }
+    debugmsg( _( "ERROR: Spell %s has invalid energy source." ), id().c_str() );
```
Comment (AMurkin) [inline src/magic.cpp]:
```suggestion
    debugmsg( "ERROR: Spell %s has invalid energy source.", id().c_str() );
```

File: src/magic.cpp [hunk 13]
```diff
@@ -195,3 +195,938 @@ bool spell_type::is_valid() const
...
+        return colorize( to_string( p.magic.available_mana() ), c_light_blue );
+    }
+    if( energy_source() == stamina_energy ) {
+        auto pair = get_hp_bar( p.stamina, p.get_stamina_max() );
+        return colorize( pair.first, pair.second );
+    }
+    if( energy_source() == hp_energy ) {
+        return "";
+    }
+    debugmsg( _( "ERROR: Spell %s has invalid energy source." ), id().c_str() );
```
Comment (AMurkin) [inline src/magic.cpp]:
```suggestion
    debugmsg( "ERROR: Spell %s has invalid energy source.", id().c_str() );
```

File: src/magic.cpp [hunk 14]
```diff
@@ -195,3 +195,938 @@ bool spell_type::is_valid() const
...
+        add_msg( m_bad, _( "All the dust in the air here falls to the ground." ) );
+    } else if( deep_pit.count( ter_here ) == 1 ) {
+        g->m.ter_set( target, t_hole );
+        add_msg( _( "The pit has deepened further." ) );
+    } else if( shallow_pit.count( ter_here ) == 1 ) {
+        g->m.ter_set( target, t_pit );
+        add_msg( _( "More debris shifts out of the pit." ) );
+    } else if( soft_dirt.count( ter_here ) == 1 ) {
+        g->m.ter_set( target, t_pit_shallow );
+        add_msg( _( "The earth moves out of the way for you" ) );
```
Comment (AMurkin) [inline src/magic.cpp]:
```suggestion
        add_msg( _( "The earth moves out of the way for you." ) );
```

File: src/magic.cpp [hunk 15]
```diff
@@ -195,3 +195,938 @@ bool spell_type::is_valid() const
...
+    damage_instance dmg;
+    dmg.add_damage( dmg_type(), damage() );
+    return dmg;
+}
+
+dealt_damage_instance spell::get_dealt_damage_instance() const
+{
+    dealt_damage_instance dmg;
+    dmg.set_damage( dmg_type(), damage() );
+    add_msg( "%i", damage() );
```
Comment (AMurkin) [inline src/magic.cpp]: Is this for debug?

File: src/activity_handlers.cpp [hunk 16]
```diff
@@ -3591,6 +3593,128 @@ void activity_handlers::hack_safe_finish( player_activity *act, player *p )
     act->set_to_null();
 }
 
+void blood_magic( player *p, int cost )
```
Comment (ZhilkinSerg) [inline src/activity_handlers.cpp]:
```suggestion
static void blood_magic( player *p, int cost )
```

File: src/panels.cpp [hunk 17]
```diff
@@ -805,6 +807,24 @@ static std::pair<nc_color, std::string> power_stat( const player &u )
     return std::make_pair( c_pwr, s_pwr );
 }
 
+std::pair<nc_color, std::string> mana_stat( const player &u )
```
Comment (ZhilkinSerg) [inline src/panels.cpp]:
```suggestion
static std::pair<nc_color, std::string> mana_stat( const player &u )
```

**Pull request merged**